agent/agent.go

```diff
@@ -0,0 +1,206 @@
+			if err := postPlugin.AfterInit(); err != nil {
+				return err
+			}
+		} else {
+			logrus.DefaultLogger().Debugf("plugin %v has no AfterInit", p)
+		}
+	}
+	a.stopCh = make(chan struct{}, 1) // If we are started, we have a stopCh to signal stopping
+	logrus.DefaultLogger().Info("Agent Started")
+	return nil
+}
+
+// Stop the Agent.  Calls close on all Plugins
+func (a *agent) Stop() error {
+	return a.stopOnce.Do(a.stop)
+}
+
+func (a *agent) stop() error {
+	if a.stopCh != nil { // Don't stop if we didn't start
+		defer close(a.stopCh)
+		// Close plugins
+		for _, p := range a.opts.Plugins {
+			if err := p.Close(); err != nil {
+				return err
+			}
+		}
+		logrus.DefaultLogger().Info("Agent Stopped.")
+		return nil
+	}
+	err := errors.New("attempted to stop an agent that wasn't Started")
+	logrus.DefaultLogger().Error(err)
+	return err
+
+}
+
+// Wait will not return until a SIGINT, SIGTERM, or SIGKILL is received
+// Or the Agent is Stopped
+// All Plugins are Closed() before Wait returns
+func (a *agent) Wait() error {
+	if a.stopCh != nil { // Don't wait if we didn't Start
+		<-a.After()
+		// If we get here, a.Stop() has already been called, and we are simply
+		// retrieving the error if any squirreled away by stopOnce
+		return a.Stop()
+	}
+	err := errors.New("attempted to wait on an agent that wasn't Started")
+	logrus.DefaultLogger().Error(err)
+	return err
+}
+
+// Run runs the agent.  Run will not return until a SIGINT, SIGTERM, or SIGKILL is received
+func (a *agent) Run() error {
+	if err := a.Start(); err != nil {
+		return err
+	}
+	return a.Wait()
+}
+
+// After returns a channel that will be closed when the agent is Stopped.
+// To retrieve any error from the agent stopping call Error() on the agent
+// The normal pattern of use is:
+//
+// agent := NewAgent(options...)
+// agent.Start()
+// select {
+// case <-agent.After() // Will wait till the agent is stopped
+// ...
+// }
+// err := agent.Error() // Will return any error from the agent being stopped
+//
+func (a *agent) After() <-chan struct{} {
+	if a.stopCh != nil {
+		return a.stopCh
+	}
+	// The agent didn't start, so we can't return a.stopCh
+	// because *only* a.start() should allocate that
+	// we won't return a nil channel, because nil channels
+	// block forever.
+	// Since the normal pattern is to call a.After() so you
+	// can select till the agent is done and a.Stop() to
+	// retrieve the error, returning a closed channel will preserve that
+	// usage, as a.Stop() returns an error complaining that the agent
+	// never started.
+	stopCh := make(chan struct{})
+	close(stopCh)
+	return stopCh
+}
+
+// Error returns any error that occurred when the agent was Stopped
+func (a *agent) Error() error {
+	// a.Stop() returns whatever error occurred when stopping the agent
+	// This is because of stopOnce
+	// If you try to retrieve an error before the agent is started, you will get
+	// an error complaining the agent isn't started.
+	return a.Stop()
```

Comment:
Interestingly, I originally did this with a.Stop().

Here's the use case.  A user uses <-a.After() in a select to wait on the agent and several other things.  a.After() comes back, and we want to retrieve the error.  It seemed unintuitive to call a.Stop() just to retrieve the error (though it would work).

This led me to thinking of other situations where one might want to retrieve the error from a Stopped agent, and so I added a.Error() to make it semantically unconfusing.